### ISSUE TITLE
xubuntu 16.04.1 X11rdp constantly eats memory

### ISSUE THREAD

**Issue by coolemza** [neutrinolabs/xrdp#408]
xubuntu 16.04.1 fresh install, xrdp build via:
sudo ./X11rdp-o-matic.sh --branch devel --justdoit

X11rdp eats memory until system kill it, i dont know is it right place to put issues about X11rdp memory leeaks...

**Comment by coolemza** [comment]
lubuntu 16.04.1 same situation, after 2 days working, X11rdp process eats 4,5 of 8Gb

**Comment by coolemza** [comment]
Debian 8.5 same sutuation

**Comment by coolemza** [comment]
cannot build devel branch with sudo ./X11rdp-o-matic.sh --branch devel --justdoit

Downloading xrdp source from the GIT repository...
Cloning into '/home/lc/tmp/X11RDP-o-Matic/work/xrdp'...
remote: Counting objects: 502, done.
remote: Compressing objects: 100% (445/445), done.
remote: Total 502 (delta 119), reused 177 (delta 44), pack-reused 0
Receiving objects: 100% (502/502), 1.90 MiB | 0 bytes/s, done.
Resolving deltas: 100% (119/119), done.
Checking connectivity... done.
patching file buildx.sh
Hunk #1 FAILED at 250.
1 out of 1 hunk FAILED -- saving rejects to file buildx.sh.rej

**Comment by metalefty** [comment]
Adding `--nocpuoptimize` option is workaround.  Now CPU optimization is merged into xrdp. More than 1 CPU cores will be used even `--nocpuoptimize` is specified. I will fix the issue shortly.

**Comment by coolemza** [comment]
yeah, just removed this patch from X11rdp-o-matic.sh and build started

**Comment by coolemza** [comment]
ubuntu 16.04 build failed with:

   dh_installdocs
cp: cannot stat 'prog_std.txt': No such file or directory
dh_installdocs: cp --reflink=auto -a prog_std.txt debian/xrdp/usr/share/doc/xrdp returned exit code 1
debian/rules:18: recipe for target 'binary' failed
make: *** [binary] Error 2
dpkg-buildpackage: error: fakeroot debian/rules binary gave error exit status 2

**Comment by coolemza** [comment]
debian 8.5 build failed with:
dpkg-source: error: cannot represent change to xorg/X11R7.6/rdp/librdp.a: binary file contents changed
dpkg-source: error: add xorg/X11R7.6/rdp/librdp.a in debian/source/include-binaries if you want to store the modified binary in the debian tarball
dpkg-source: error: cannot represent change to xorg/X11R7.6/rdp/rdpinput.o: binary file contents changed
dpkg-source: error: add xorg/X11R7.6/rdp/rdpinput.o in debian/source/include-binaries if you want to store the modified binary in the debian tarball
dpkg-source: error: unrepresentable changes to source
dpkg-buildpackage: error: dpkg-source -b xrdp-0.9.0+devel gave error exit status 2

**Comment by metalefty** [comment]
Can you try my fork of X11RDP-o-Matic?  This will be bran-new X11RDP-o-Matic.
This is still testing but probably can build current xrdp devel head.

```
git clone --branch rootless https://github.com/metalefty/X11RDP-o-Matic.git
cd X11RDP-o-Matic
./X11RDP-o-Matic --branch devel
```

**Comment by coolemza** [comment]
wow, man, you make my day, this version works like a charm, no memory leaks, no artifacts, will continue testing

**Comment by tricksol** [comment]
Using the new rootless branch I am getting this.

/build.log
patching file sesman/sesman.ini
patching file xorgxrdp/xrdpdev/xorg.conf
patching file xrdp/xrdp.ini
Hunk #1 FAILED at 144.
1 out of 1 hunk FAILED
dpkg-source: info: the patch has fuzz which is not allowed, or is malformed
dpkg-source: info: if patch 'config.diff' is correctly applied by quilt, use 'quilt refresh' to update it
dpkg-source: info: restoring quilt backup files for config.diff
dpkg-source: error: LC_ALL=C patch -t -F 0 -N -p1 -u -V never -E -b -B .pc/config.diff/ --reject-file=- < xrdp-0.9.0~20161102+gitcab8955+devel/debian/patches/config.diff gave error exit status 1
dpkg-buildpackage: error: dpkg-source --before-build xrdp-0.9.0~20161102+gitcab8955+devel gave error exit status 2

**Comment by metalefty** [comment]
I fixed it. Please try devel branch of scarygliders/X11RDP-o-Matic now.

**Comment by tricksol** [comment]
Still same error

mmidgett@bk3:~$ rm -rf X11RDP-o-Matic/
mmidgett@bk3:~$ git clone https://github.com/scarygliders/X11RDP-o-Matic.git
Cloning into 'X11RDP-o-Matic'...
remote: Counting objects: 940, done.
remote: Compressing objects: 100% (18/18), done.
remote: Total 940 (delta 7), reused 0 (delta 0), pack-reused 922
Receiving objects: 100% (940/940), 248.83 KiB | 0 bytes/s, done.
Resolving deltas: 100% (561/561), done.
Checking connectivity... done.
mmidgett@bk3:~$ cd X11RDP-o-Matic/
mmidgett@bk3:~/X11RDP-o-Matic$ sudo ./X11rdp-o-matic.sh --justdoit --branch devel

Okay, will just do the install from start to finish with no user interaction...
----------------------------------------------------------------------
----------------------------------------------------------------------
Obtaining list of available branches...
----------------------------------------------------------------------
devel gh-pages itamarjp-patch-1 master noorders v0.6 v0.7 v0.8 v0.8uni
----------------------------------------------------------------------
Using branch ==>> devel <<==
Note : using the bleeding-edge version may result in problems :)
----------------------------------------------------------------------
Using the following xrdp configuration : --prefix=/usr
----------------------------------------------------------------------
Debian package version number will be : 0.9.0+devel
----------------------------------------------------------------------
 *** Will remove the contents of /opt/X11rdp and /home/mmidgett/X11RDP-o-Matic/work/xrdp-0.9.0+devel ***

running apt-get update
Checking for previously generated packages...
----------------------------------------------------------------------
Downloading xrdp source from the GIT repository...
Cloning into '/home/mmidgett/X11RDP-o-Matic/work/xrdp'...
remote: Counting objects: 502, done.
remote: Compressing objects: 100% (445/445), done.
remote: Total 502 (delta 119), reused 175 (delta 44), pack-reused 0
Receiving objects: 100% (502/502), 1.90 MiB | 404.00 KiB/s, done.
Resolving deltas: 100% (119/119), done.
Checking connectivity... done.
patching file buildx.sh
Hunk #1 FAILED at 250.
1 out of 1 hunk FAILED -- saving rejects to file buildx.sh.rej

**Comment by metalefty** [comment]
I can't see you're using devel branch.

**Comment by tricksol** [comment]
mmidgett@bk3:~/X11RDP-o-Matic$ sudo ./X11rdp-o-matic.sh --justdoit --branch devel

Okay, will just do the install from start to finish with no user interaction...

Obtaining list of available branches...

devel gh-pages itamarjp-patch-1 master noorders v0.6 v0.7 v0.8 v0.8uni

Using branch ==>> devel <<==

Am I doing it wrong?

**Comment by metalefty** [comment]
**devel branch of scarygliders/X11RDP-o-Matic**

```
mmidgett@bk3:~$ git clone https://github.com/scarygliders/X11RDP-o-Matic.git
```
`git checkout devel` after this will work for you :)

**Comment by coolemza** [comment]
after rebuilding with devel scarygliders branch, xrdp stopped working for me:

lc@work:~$ sudo journalctl -u xrdp

ноя 15 17:06:59 work systemd[1]: Starting xrdp (X Remote Desktop Protocol) protocol server...
ноя 15 17:06:59 work systemd[1]: xrdp.service: PID file /run/xrdp/xrdp.pid not readable (yet?) after start: No such file or directory
ноя 15 17:07:00 work systemd[1]: Started xrdp (X Remote Desktop Protocol) protocol server.
ноя 15 17:07:01 work xrdp[840]: (840)(140014509463296)[INFO ] starting xrdp with pid 840
ноя 15 17:07:01 work xrdp[840]: (840)(140014509463296)[INFO ] listening to port 3389 on 0.0.0.0
ноя 15 17:07:37 work xrdp[840]: (840)(140014509463296)[INFO ] A connection received from: 192.168.122.1 port 45146
ноя 15 17:07:37 work xrdp[840]: (840)(140014509463296)[DEBUG] Closed socket 12 (AF_INET 192.168.122.160:3389)
ноя 15 17:07:37 work xrdp[1188]: (1188)(140014509463296)[DEBUG] Closed socket 11 (AF_INET 0.0.0.0:3389)
ноя 15 17:07:37 work xrdp[1188]: (1188)(140014509463296)[CORE ] Warning: Your configured security layer is undefined, xrdp will negotiate client compatible
ноя 15 17:07:37 work xrdp[1188]: (1188)(140014509463296)[INFO ] Using default X.509 certificate: /etc/xrdp/cert.pem
ноя 15 17:07:37 work xrdp[1188]: (1188)(140014509463296)[INFO ] Using default X.509 key file: /etc/xrdp/key.pem
ноя 15 17:07:38 work xrdp[1188]: (1188)(140014509463296)[DEBUG] xrdp_000004a4_wm_login_mode_event_00000001
ноя 15 17:07:38 work xrdp[1188]: (1188)(140014509463296)[INFO ] Loading keymap file /etc/xrdp/km-00000409.ini
ноя 15 17:07:38 work xrdp[1188]: (1188)(140014509463296)[WARN ] local keymap file for 0x00000409 found and doesn't match built in keymap, using local keymap file
ноя 15 17:07:38 work xrdp[1188]: (1188)(140014509463296)[DEBUG] xrdp_wm_log_msg: connecting to sesman ip 127.0.0.1 port 3350
ноя 15 17:07:38 work xrdp[1188]: (1188)(140014509463296)[INFO ] xrdp_wm_log_msg: sesman connect ok
ноя 15 17:07:38 work xrdp[1188]: (1188)(140014509463296)[DEBUG] xrdp_wm_log_msg: sending login info to session manager, please wait...
ноя 15 17:07:38 work xrdp[1188]: (1188)(140014509463296)[DEBUG] return value from xrdp_mm_connect 0
ноя 15 17:07:39 work xrdp[1188]: (1188)(140014509463296)[INFO ] xrdp_wm_log_msg: login successful for display 10
ноя 15 17:07:39 work xrdp[1188]: (1188)(140014509463296)[DEBUG] xrdp_wm_log_msg: started connecting
ноя 15 17:07:42 work xrdp[1188]: (1188)(140014509463296)[DEBUG] Closed socket 17 (AF_UNIX)
ноя 15 17:07:46 work xrdp[1188]: (1188)(140014509463296)[DEBUG] Closed socket 17 (AF_UNIX)
ноя 15 17:07:49 work xrdp[1188]: (1188)(140014509463296)[DEBUG] Closed socket 17 (AF_UNIX)
ноя 15 17:07:53 work xrdp[1188]: (1188)(140014509463296)[DEBUG] Closed socket 17 (AF_UNIX)
....

ноя 15 17:11:09 work xrdp[1188]: (1188)(140014509463296)[DEBUG] xrdp_wm_log_msg: connection problem, giving up
ноя 15 17:11:09 work xrdp[1188]: (1188)(140014509463296)[DEBUG] Closed socket 17 (AF_UNIX)
ноя 15 17:11:09 work xrdp[1188]: (1188)(140014509463296)[DEBUG] xrdp_wm_log_msg: some problem
ноя 15 17:11:09 work xrdp[1188]: (1188)(140014509463296)[DEBUG] xrdp_mm_module_cleanup
ноя 15 17:11:09 work xrdp[1188]: (1188)(140014509463296)[DEBUG] Closed socket 16 (AF_INET 127.0.0.1:60054)
ноя 15 17:11:09 work xrdp[1188]: (1188)(140014509463296)[DEBUG] Closed socket 12 (AF_INET 192.168.122.160:3389)
ноя 15 17:11:09 work xrdp[1188]: (1188)(140014509463296)[DEBUG] xrdp_mm_module_cleanup
ноя 15 17:11:09 work xrdp[1188]: (1188)(140014509463296)[ERROR] Listening socket is in wrong state, terminating listener

**Comment by coolemza** [comment]
also build script cannot find file:

Do the following command to see build progress.
	$ tail -f /tmp/tmp.VcQSt6QHsI.X11RDP-o-Matic/build.log
Using the following xrdp configuration: 
	--prefix=/usr --sysconfdir=/etc --localstatedir=/var --enable-fuse --enable-jpeg --enable-opus
sed: can't read /tmp/tmp.VcQSt6QHsI.X11RDP-o-Matic/xorg/X11R7.6/buildx.sh: No such file or directory
Removing installed x11rdp... 
Following command will be executed via sudo:
	apt-get -y remove x11rdp
done

Following command will be executed via sudo:
	rm -rf /opt/X11rdp

Following command will be executed via sudo:
	mkdir -p /opt/X11rdp

**Comment by proski** [comment]
I ran X11rdp under valgrind this weekend, but I haven't seen any major memory leaks. Just a few kilobytes here and there. Likewise, my fixes to xrdp are fairy minor. I suggest running X11rdp under valgrind on an affected system. Use a shell wrapper, as the display number (e.g. `:10`) is inserted after the first parameter. Make sure to use unstripped X11rdp (buildx.sh fixed two days ago to only strip the installed binary).

**Comment by coolemza** [comment]
Yes, there was no memory leaks at least major, but when i tryed to move to 'devel' branch of scarygliders. It builded but i cannot connect to xrdp using remmina, i see only black screen, log i posted above, so i decided to build 'rootless' branch from metalefty fork, and build failed with:

**Comment by coolemza** [comment]
upgraded to xubuntu 16.10, same situation, 'rootless' branch build failed, 'devel' branch build ok, but connection failed with the same logs

**Comment by metalefty** [comment]
Here is not a place to debug o-Matic. The original issue should be solved in v0.9.2. 
Please test v0.9.2 with xorgxrdp. If the issue still occurs with v0.9.2, please reopen.

**Comment by coolemza** [comment]
yes its solved in 0.9.2
